cpp/src/io/utilities/config_utils.cpp

```diff
@@ -87,38 +89,171 @@ bool is_stable_enabled() { return is_all_enabled() or get_env_policy() == usage_
+  }
+
+  void* do_allocate_async(std::size_t bytes, std::size_t alignment, cuda::stream_ref stream)
+  {
+    if (bytes <= pool_size_) {
+      try {
+        return pool_->allocate_async(bytes, alignment, stream);
+      } catch (...) {
+        // If the pool is exhausted, fall back to the upstream memory resource
+      }
+    }
+
+    return upstream_mr_.allocate_async(bytes, alignment, stream);
+  }
+
+  void do_deallocate_async(void* ptr,
+                           std::size_t bytes,
+                           std::size_t alignment,
+                           cuda::stream_ref stream) noexcept
+  {
+    if (bytes <= pool_size_ && ptr >= pool_begin_ && ptr <= pool_end_) {
+      pool_->deallocate_async(ptr, bytes, alignment, stream);
+    } else {
+      upstream_mr_.deallocate_async(ptr, bytes, alignment, stream);
+    }
+  }
+
+  void* allocate_async(std::size_t bytes, cuda::stream_ref stream)
+  {
+    return do_allocate_async(bytes, rmm::RMM_DEFAULT_HOST_ALIGNMENT, stream);
+  }
+
+  void* allocate_async(std::size_t bytes, std::size_t alignment, cuda::stream_ref stream)
+  {
+    return do_allocate_async(bytes, alignment, stream);
+  }
+
+  void* allocate(std::size_t bytes, std::size_t alignment = rmm::RMM_DEFAULT_HOST_ALIGNMENT)
+  {
+    auto const result = do_allocate_async(bytes, alignment, stream_);
+    stream_.wait();
+    return result;
+  }
+
+  void deallocate_async(void* ptr, std::size_t bytes, cuda::stream_ref stream) noexcept
+  {
+    return do_deallocate_async(ptr, bytes, rmm::RMM_DEFAULT_HOST_ALIGNMENT, stream);
+  }
+
+  void deallocate_async(void* ptr,
+                        std::size_t bytes,
+                        std::size_t alignment,
+                        cuda::stream_ref stream) noexcept
+  {
+    return do_deallocate_async(ptr, bytes, alignment, stream);
+  }
+
+  void deallocate(void* ptr,
+                  std::size_t bytes,
+                  std::size_t alignment = rmm::RMM_DEFAULT_HOST_ALIGNMENT) noexcept
+  {
+    deallocate_async(ptr, bytes, alignment, stream_);
+    stream_.wait();
+  }
+
+  bool operator==(fixed_pinned_pool_memory_resource const& other) const
+  {
+    return pool_ == other.pool_ and stream_ == other.stream_;
+  }
+
+  bool operator!=(fixed_pinned_pool_memory_resource const& other) const
+  {
+    return !operator==(other);
+  }
+
+  [[maybe_unused]] friend void get_property(fixed_pinned_pool_memory_resource const&,
+                                            cuda::mr::device_accessible) noexcept
+  {
+  }
+
+  [[maybe_unused]] friend void get_property(fixed_pinned_pool_memory_resource const&,
+                                            cuda::mr::host_accessible) noexcept
+  {
+  }
+};
+
+static_assert(cuda::mr::resource_with<fixed_pinned_pool_memory_resource,
+                                      cuda::mr::device_accessible,
+                                      cuda::mr::host_accessible>,
+              "");
+
+rmm::host_async_resource_ref default_pinned_mr()
```

Comment:
We've been using this API like this:

1. We create a pinned pool that is larger (few GB) that is used for our java allocations as well
2. Then we call `set_host_memory_resource(our_pool)`. 
3. This new code will allocate another pool of up to 100MB pinned that won't be used by us, since we are calling the set method with a new pool.
4. When we are shutting down, we rollback this change and call `set_host_memory_resource` with the default resource we got from you from the set call.

100MB is not a lot, but it would be nice to skip allocating this if it's not needed. Also, environment variables are not great for us to be able to configure things from spark-rapids.

Another issue with the static allocation here is that I have to call a `get_host_memory_resource` or `set_host_memory_resource` in order to initialize the pool. To me it seems we need a `configure_host_memory_resource` function, which allows us to pass a pool size as an optional or with 0 as a special value. This function would return a resource, either pooled or the thin wrapper around `cudaMallocHost`/`cudaFreeHost`, this thin wrapper would be the "default resource" if the configured size is 0.

Additionally, if you want to have the default pinned pool for cuDF then the default value for the pool size could be > 0, as you have here. Then the callers of `get_host_memory_resource` work as you intended.

Reply:
There is a race condition with the above. We'd have to at least document really well that any "configuration" function must be called before any cuDF api.